### PR TITLE
Add `content_for` tag parser support + `ValidContentForArguments` check

### LINKED ISSUE #466: The `content_for` tag should be supported by the parser and parse its arguments.
**Describe the bug**
Right now any kind of `{% content_for 'block', id: something, name: something %}` parses the markup as a flat string. It should be something easier to manage in checks/complete/etc instead. 

https://shopify.dev/docs/api/liquid/tags#content_for

**Source**

```liquid
{% content_for 'block', type: "", id: "" %}
```

**Expected behaviour**
```ts
const ast: LiquidHtmlNode = toLiquidHtmlAST(`{% content_for 'block', type: "", id: "" %}`);
const tag = ast.children[0];
assert(tag.name === 'content_for');
assert(tag.markup.type === 'ContentForMarkup');
// something like that
assert(tag.markup.contentForType  === 'block');
assert(tag.markup.kwargs.type === '...');
```

**Actual behaviour**
```ts
const ast: LiquidHtmlNode = toLiquidHtmlAST(`{% content_for 'block', type: "", id: "" %}`);
const tag = ast.children[0];
assert(tag.name === 'content_for');
// markup isn't a complex type, but a string! It means we need to do regex!
assert(typeof tag.markup === 'string');
```

**Additional context**
- I think we should "inspire ourselves" by [how we parse the `render` tag][1]
- [this is how we describe the ConcreteLiquidTagRenderMarkup type](https://github.com/Shopify/theme-tools/blob/main/packages/liquid-html-parser/src/stage-1-cst.ts#L329-L335)
- [this is where we map from the positional ohm arguments to the type](https://github.com/Shopify/theme-tools/blob/main/packages/liquid-html-parser/src/stage-1-cst.ts#L778-L787)
- [this is where we define the RenderMarkup interface](https://github.com/Shopify/theme-tools/blob/main/packages/liquid-html-parser/src/stage-2-ast.ts#L380-L390)
- [this is where we turn a ConcreteLiquidTagRenderMarkup into a RenderMarkup](https://github.com/Shopify/theme-tools/blob/main/packages/liquid-html-parser/src/stage-2-ast.ts#L1683-L1693)

At the end of the day, we should be able to do "richer" completion and theme checks when it comes to that tag once we parse it better. 

[1]: https://github.com/Shopify/theme-tools/blob/main/packages/liquid-html-parser/grammar/liquid-html.ohm#L129-L134

### PR DESCRIPTION
**Pull request by charlespwd**
Fixes #466

## What are you adding in this PR?

- Add parser support for the `content_for` tag
- Updated the old `content_for` checks to use the AST rather than regexes to perform their work
- Add prettier plugin support for the `content_for` tag
- Add the `ValidContentForArguments` check
  - For `{% content_for "blocks" %}`
    - Only `context.*` arguments are accepted
  - For `{% content_for "block" %}`
    - `type` and `id` are required
    - `type` and `id` must be strings
    - `context.*` arguments are optional
    - nothing else is accepted

## What's next? Any followup issues?

- [ ] Document `ValidContentForArguments` on shopify.dev
- [x] #569 
- [x] #570
- [x] #571 

## What did you learn?

There were more things I needed to change than I remembered. TypeScript helps a ton here because it complains any time you add a new node type.

## Before you deploy

- [x] This PR includes a new checks or changes the configuration of a check
  - [x] I included a minor bump `changeset`
  - [x] It's in the `allChecks` array in `src/checks/index.ts`
  - [x] I ran `yarn build` and committed the updated configuration files
- [x] I included a minor bump `changeset`
- [x] My feature is backward compatible

### CODE REVIEW

File: packages/prettier-plugin-liquid/src/printer/print/liquid.ts [hunk 1]
```diff
@@ -165,6 +165,12 @@ function printNamedLiquidBlockStart(
       ]);
     }
 
+    case NamedTags.content_for: {
+      const markup = node.markup;
+      const trailingWhitespace = markup.args.length > 0 ? line : ' ';
+      return tag(trailingWhitespace);
+    }
```
Comment (charlespwd) [inline packages/prettier-plugin-liquid/src/printer/print/liquid.ts:172]: The `tag(trailingWhitespace)` helper is here to handle the logic about what "whitespace-logic-char" we want to use at the end of the tag.

```
// when it doesn't break
tag('') => {% content_for "blocks"%}
tag(' ') => {% content_for "blocks" %}
tag(line) => {% content_for "blocks" %}

// when it breaks
tag('') => {% content_for "blocks"%}
tag(' ') => {% content_for "blocks" %}
tag(line) => {% content_for "blocks"
%}
```

**Review by aswamy — APPROVED** [review]
small stuff

File: .changeset/tasty-mugs-behave.md [hunk 2]
```diff
@@ -0,0 +1,7 @@
+---
+'@shopify/liquid-html-parser': minor
+'@shopify/prettier-plugin-liquid': minor
+'@shopify/theme-language-server-common': minor
+---
+
+Add support for the `content_for` Liquid tag
```
[the remark this answers is not in the document]
Comment (charlespwd) [reply .changeset/tasty-mugs-behave.md:7]: I didn't want three changesets, one for the prettier plugin, one for the parser and one for language server.

Support means something different for each one of them.

File: packages/liquid-html-parser/src/stage-1-cst.spec.ts [hunk 3]
```diff
@@ -586,6 +586,45 @@ describe('Unit: Stage 1 (CST)', () => {
...
+
+      it('should parse content_for "block", id: "my-id", type: "my-block"', () => {
+        for (const { toCST, expectPath } of testCases) {
+          cst = toCST(`{% content_for "block", id: "block-id", type: "block-type" -%}`);
+          expectPath(cst, '0.type').to.equal('LiquidTag');
+          expectPath(cst, '0.name').to.equal('content_for');
+          expectPath(cst, '0.markup.type').to.equal('ContentForMarkup');
+          expectPath(cst, '0.markup.contentForType.type').to.equal('String');
+          expectPath(cst, '0.markup.contentForType.value').to.equal('block');
+          expectPath(cst, '0.markup.contentForType.single').to.equal(false);
```
[the remark this answers is not in the document]
Comment (charlespwd) [reply packages/liquid-html-parser/src/stage-1-cst.spec.ts:613]: Single is for single or double quote

File: packages/theme-check-common/src/checks/unique-static-block-id/index.ts [hunk 4]
```diff
@@ -21,41 +23,36 @@ export const UniqueStaticBlockId: LiquidCheckDefinition = {
...
-        if (usedIds.has(filteredIdValue)) {
-          const nodeInSource = node.source.substring(node.position.start);
-          const contentForBlockStartIndex = nodeInSource.indexOf(blockType);
-
-          const idParamIndex = idValueMatch.index + contentForBlockStartIndex;
-          const idParamValueLength = entireIdTerm.length;
+        const idValueNode = idNode.value;
+        if (idValueNode.type !== NodeTypes.String) {
+          return; // covered by VariableContentForArguments
+        }
```
[the remark this answers is not in the document]
Comment (charlespwd) [reply packages/theme-check-common/src/checks/unique-static-block-id/index.ts:43]: Asked internally and this is not accepted/rejected by the backend right now

**Pull request merged**